src/Storages/StorageKeeperMap.cpp

```diff
@@ -473,20 +473,33 @@ bool StorageKeeperMap::dropTable(zkutil::ZooKeeperPtr zookeeper, const zkutil::E
         case ZOK:
         {
             metadata_drop_lock->setAlreadyRemoved();
-            completely_removed = true;
+            drop_done = true;
             LOG_INFO(log, "Metadata ({}) and data ({}) was successfully removed from ZooKeeper", metadata_path, data_path);
             break;
         }
         case ZNONODE:
             throw Exception(ErrorCodes::LOGICAL_ERROR, "There is a race condition between creation and removal of metadata. It's a bug");
         case ZNOTEMPTY:
-            LOG_ERROR(log, "Metadata was not completely removed from ZooKeeper");
+        {
+            // valid case when this can happen is if a table checked "dropped" path just before it was created.
+            // new table will create data/metadata paths again while drop is in progress
+            // only bad thing that can happen is if we start inserting data into new table while
+            // we remove data here (some data can be lost)
+            LOG_WARNING(log, "Metadata was not completely removed from ZooKeeper. Maybe some other table is using the same path");
+
+            // we need to remove at least "dropped" nodes
+            Coordination::Requests requests;
+            ops.emplace_back(zkutil::makeRemoveRequest(metadata_drop_lock->getPath(), -1));
+            ops.emplace_back(zkutil::makeRemoveRequest(dropped_path, -1));
+            zookeeper->multi(requests);
```

Comment:
I'm not sure that it's good idea, it would be better to retry/throw exception if "dropped" node exists on table creation

Reply:
that's what it does now, it tries to get a lock for dropping, if it fails it will retry

the problem is the following situation:
table 1 is created

table 2 is created
table 2 checks if there is "dropped", doesn't find anything

table 1 is dropped
table 1 creates "dropped", and removes "tables"

table 2 create "tables"

table 1 now can't remove "dropped" with "metadata" because "metadata will contain recreated "tables" path

Reply:
Is it possible to make table creation and check for "dropped" atomic (i.e. in one zk-transaction)?

Reply:
With current ZK operations, no.
Multiple problems:
I can't check that something doesn't exist, anything that isn't ZOK will fail the transaction.
I need to create multiple nodes (metadata and tables) that can exist, but it's possible one or both don't.

That would be the best option, but I don't see how to make it work.
If you have some idea, that would be great.

Reply:
`StorageReplicatedMergeTree` does something similar (see `createTableIfNotExists`, `createReplica`, `dropReplica` and `01305_replica_create_drop_zookeeper`), but it's quite complicated.

 Where is the "commit point" of KeeperMap replica creation? Seems like it's here:
https://github.com/ClickHouse/ClickHouse/blob/deee0d639fe479f00cdda11830182350e8e340cf/src/Storages/StorageKeeperMap.cpp#L370
(however, there's another issue: `data_path` will not be created if connection loss happened after crating `table_path`, because `tryCreate(table_path, ...)` will return `ZNODEEXISTS` on retry and exception will be thrown, so it's not really a commit point)

We can create `table_path` and check that `dropped_path` does not exists in a transaction, so it will not be possible to add new replica when table is being dropped. If such transaction fails, we will retry, finish removal of leftovers and create table metadata.

Also it makes sense to split creation into two operations: create a table (`metadata_path`, `data_path`, `tables_path`) and add new replica `table_path`, so it will be possible to create all nodes in one transaction and don't care if some nodes already exist (we will simply "add new replica" (atomically with check that table metadata was not dropped/changed/recreated) if table nodes already exist).

P.S. ZooKeeper API does not allow to check if some node does not exists, but it can be done with this trick:
https://github.com/ClickHouse/ClickHouse/blob/acde29e314e4d1f66fe5a08984c712a75e8e66fd/src/Storages/StorageReplicatedMergeTree.cpp#L1401-L1403

Reply:
Thanks, it's much, much clearer now.
I checked `StorageReplicatedMergeTree` a bunch of times, but now I think I get the idea completely.

I'll make the PR active again once I think it's okay.
And the "not exist" check is really nice!